dttools/src/jx.h

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright (C) 2015- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#ifndef JX_H
+#define JX_H
+
+/** @file jx.h JSON Expressions (JX) library.
+This module implements extended JSON expressions as a C library.
+We use our own custom library to avoid external dependencies and
+to implement various extensions to strict JSON.
+
+For example, the following bit of code:
+
+<pre>
+struct jx * obj = jx_object(
+	jx_pair(
+		jx_string("hello"),
+		jx_string("world"),
+		0)
+	);
+
+jx_print_file(jx,stdout);
+
+jx_delete(jx);
+</pre>
+
+Will create the following output:
+<pre>
+{ "hello" : "world" }
+</pre>
+
+@see jx_parse.h
+@see jx_print.h
+*/
+
+#include <stdint.h>
+
+/** JX atomic type.  */
+
+typedef enum {
+	JX_NULL,	/**< null value */
+	JX_BOOLEAN,	/**< true or false */
+	JX_INTEGER,	/**< integer value */
+	JX_DOUBLE,	/**< floating point value */
+	JX_STRING,	/**< string value */
+	JX_SYMBOL,	/**< variable identifier */
+	JX_ARRAY,	/**< array containing values */
+	JX_OBJECT,	/**< object containing key-value pairs */
+} jx_type_t;
+
+typedef int64_t jx_int_t;
+
+/** JX item linked-list used by @ref JX_ARRAY and @ref jx.items */
+
+struct jx_item {
+	struct jx      *value;	/**< value of this item */
+	struct jx_item *next;	/**< pointer to next item */
+};
+
+/** JX key-value pairs used by @ref JX_OBJECT and @ref jx.pairs */
+
+struct jx_pair {
+	struct jx      *key;	/**< key of this pair */
+	struct jx      *value;  /**< value of this pair */
+	struct jx_pair *next;   /**< pointer to next pair */
+};
+
+/** JX value representing any expression type. */
+
+struct jx {
+	jx_type_t type;               /**< type of this value */
+	union {
+		int boolean_value;      /**< value of @ref JX_BOOLEAN */
+		jx_int_t integer_value; /**< value of @ref JX_INTEGER */
+		double double_value;   /**< value of @ref JX_DOUBLE */
+		char * string_value;  /**< value of @ref JX_STRING */
+		char * symbol_name;   /**< value of @ref JX_SYMBOL */
+		struct jx_item *items;  /**< value of @ref JX_ARRAY */
+		struct jx_pair *pairs;  /**< value of @ref JX_OBJECT */
+	};
+};
```

Comment:
They could be, but that would prevent code from traversing the data structure.  For example, `catalog_query` loads an entire object and then iterates over the pairs one by one.